### PR TITLE
Gauge API : Add LongGauge Support (Part1)

### PR DESCRIPTION
**Pull request by mayurkale22**
These APIs will allow users to report already aggregated data into Open Census.

**Part1 : Add LongGauge Support**.
Part2 : Add DoubleGauge Support.
Part3 : Add DerivedLongGauge Support.
Part4 : Add DerivedDoubleGauge Support.

### CODE REVIEW

File: api/src/main/java/io/opencensus/metrics/LongGauge.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+package io.opencensus.metrics;
+
+import io.opencensus.internal.Utils;
+import java.util.List;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Long Gauge metric, to report instantaneous measurement of an int64 value. Gauges can go both up
+ * and down. The gauges values can be negative. For example, the number of pending jobs in the
+ * queue. See {@link io.opencensus.metrics.MetricRegistry} for an example of its use.
```
Comment (bogdandrutu) [inline api/src/main/java/io/opencensus/metrics/LongGauge.java]: I think the example is here, am I wrong?

File: api/src/main/java/io/opencensus/metrics/LongGauge.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+package io.opencensus.metrics;
+
+import io.opencensus.internal.Utils;
+import java.util.List;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Long Gauge metric, to report instantaneous measurement of an int64 value. Gauges can go both up
+ * and down. The gauges values can be negative. For example, the number of pending jobs in the
```
Comment (bogdandrutu) [inline api/src/main/java/io/opencensus/metrics/LongGauge.java]: Remove the "For example, the number of pending jobs in the queue". You have a real example for that.

File: api/src/main/java/io/opencensus/metrics/LongGauge.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+ * class YourClass {
+ *
+ *   private static final MetricRegistry metricRegistry = Metrics.getMetricRegistry();
+ *
+ *   List<LabelKey> labelKeys = Arrays.asList(LabelKey.create("Name", "desc"));
+ *   List<LabelValue> labelValues = Arrays.asList(LabelValue.create("Inbound"));
+ *
+ *   LongGauge gauge = metricRegistry.addLongGauge("queue_size", "Pending jobs", "1", labelKeys);
+ *
+ *   // It is strongly recommended to keep a reference of a point for manual operations.
```
Comment (bogdandrutu) [inline api/src/main/java/io/opencensus/metrics/LongGauge.java]: ditto.

File: api/src/main/java/io/opencensus/metrics/LongGauge.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+ *
+ * <pre>{@code
+ * class YourClass {
+ *
+ *   private static final MetricRegistry metricRegistry = Metrics.getMetricRegistry();
+ *
+ *   List<LabelKey> labelKeys = Arrays.asList(LabelKey.create("Name", "desc"));
+ *   LongGauge gauge = metricRegistry.addLongGauge("queue_size", "Pending jobs", "1", labelKeys);
+ *
+ *   // It is strongly recommended to keep a reference of a point for manual operations.
```
Comment (bogdandrutu) [inline api/src/main/java/io/opencensus/metrics/LongGauge.java]: I would remove "strongly" but everything else sounds good.

File: api/src/main/java/io/opencensus/metrics/LongGauge.java [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+ * }
+ * }</pre>
+ *
+ * @since 0.17
+ */
+@ThreadSafe
+public abstract class LongGauge {
+
+  /**
+   * Creates a {@code TimeSeries} and returns a {@code LongPoint}, which is part of the TimeSeries.
```
Comment (bogdandrutu) [inline api/src/main/java/io/opencensus/metrics/LongGauge.java]: You don't always create here so the comment should reflect that.

File: api/src/main/java/io/opencensus/metrics/LongGauge.java [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+   * This is more convenient form when you want to manually increase and decrease values as per your
```
Comment (bogdandrutu) [inline api/src/main/java/io/opencensus/metrics/LongGauge.java]: "This is more convenient form when you want to manually increase and decrease values as per your service requirements" this sounds more like a class comment or a comment in the MetricRegistry.

File: api/src/main/java/io/opencensus/metrics/LongGauge.java [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+   * service requirements. The number of label values must be the same to that of the label keys
+   * passed to {@link MetricRegistry#addLongGauge}.
+   *
+   * <p>It is strongly recommended to keep a reference to the LongPoint instead of always calling
```
Comment (bogdandrutu) [inline api/src/main/java/io/opencensus/metrics/LongGauge.java]: ditto. remove strongly.

File: api/src/main/java/io/opencensus/metrics/LongGauge.java [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    NoopLongGauge(String name, String description, String unit, List<LabelKey> labelKeys) {
+      Utils.checkNotNull(name, "name");
+      Utils.checkNotNull(description, "description");
+      Utils.checkNotNull(unit, "unit");
+      Utils.checkNotNull(labelKeys, "labelKeys should not be null.");
+      Utils.checkListElementNotNull(labelKeys, "labelKeys element should not be null.");
+    }
+
+    @Override
+    public NoopLongPoint getOrCreateTimeSeries(List<LabelValue> labelValues) {
```
Comment (bogdandrutu) [inline api/src/main/java/io/opencensus/metrics/LongGauge.java]: Check for the size of label values here. To do that you can have an int with the size of the label keys.

File: api/src/main/java/io/opencensus/metrics/LongGauge.java [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+   * This is more convenient form when you want to manually increase and decrease values as per your
+   * service requirements. The number of label values must be the same to that of the label keys
+   * passed to {@link MetricRegistry#addLongGauge}.
+   *
+   * <p>It is strongly recommended to keep a reference to the LongPoint instead of always calling
+   * this method for manual operations.
+   *
+   * @param labelValues the list of label values.
+   * @return a {@code LongPoint} the value of single gauge.
+   * @throws NullPointerException if {@code labelValues} is null OR element of {@code labelValues}
```
Comment (bogdandrutu) [inline api/src/main/java/io/opencensus/metrics/LongGauge.java]: OR **any** element?

File: api/src/test/java/io/opencensus/metrics/MetricRegistryTest.java [hunk 10]
```diff
@@ -33,153 +32,39 @@
...
-        });
+    metricRegistry.addLongGauge("name", "description", "1", null);
   }
 
   @Test
-  public void addLongGauge_NullFunction() {
-    thrown.expect(NullPointerException.class);
-    metricRegistry.addLongGauge(
-        "name", "description", "1", new LinkedHashMap<LabelKey, LabelValue>(), null, null);
+  public void addLongGauge_NoopPoint() {
```
Comment (bogdandrutu) [inline api/src/test/java/io/opencensus/metrics/MetricRegistryTest.java]: Add tests for Null elements in getOrCreateTimeSeries and invalid size.

File: api/src/test/java/io/opencensus/metrics/MetricRegistryTest.java [hunk 11]
```diff
@@ -33,153 +32,39 @@
...
+    LongGauge longGaugeMetric =
+        metricRegistry.addLongGauge("name", "description", "1", new ArrayList<LabelKey>());
+    LongPoint dp = longGaugeMetric.getOrCreateTimeSeries(new ArrayList<LabelValue>());
+    dp.add(12);
+    dp.set(12);
+
+    longGaugeMetric.getDefaultTimeSeries();
+    longGaugeMetric.getOrCreateTimeSeries(new ArrayList<LabelValue>());
```
Comment (bogdandrutu) [inline api/src/test/java/io/opencensus/metrics/MetricRegistryTest.java]: We should have tests for remove (at least for null elements and the fact that it does not break if arguments are correct).

File: api/src/test/java/io/opencensus/metrics/MetricRegistryTest.java [hunk 12]
```diff
@@ -33,153 +32,39 @@
...
-    metricRegistry.addLongGauge(
-        "name", "description", "1", new LinkedHashMap<LabelKey, LabelValue>(), null, null);
+  public void addLongGauge_NoopPoint() {
+    LongGauge longGaugeMetric =
+        metricRegistry.addLongGauge("name", "description", "1", new ArrayList<LabelKey>());
+    LongPoint dp = longGaugeMetric.getOrCreateTimeSeries(new ArrayList<LabelValue>());
+    dp.add(12);
+    dp.set(12);
+
+    longGaugeMetric.getDefaultTimeSeries();
```
Comment (bogdandrutu) [inline api/src/test/java/io/opencensus/metrics/MetricRegistryTest.java]: Consider to move this in a separate test for the defaultTimeSeries

File: impl_core/src/main/java/io/opencensus/implcore/internal/Utils.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+  private Utils() {}
+
+  /**
+   * Throws a {@link NullPointerException} if any of the list elements is null.
+   *
+   * @param list the argument list to check for null.
+   * @param errorMessage the message to use for the exception. Will be converted to a string using
+   *     {@link String#valueOf(Object)}.
+   */
+  public static <T> void checkListElementNotNull(
```
Comment (bogdandrutu) [inline impl_core/src/main/java/io/opencensus/implcore/internal/Utils.java]: I think some tests would be good for this class.

File: impl_core/src/main/java/io/opencensus/implcore/metrics/LongGaugeImpl.java [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+/** Implementation of {@link LongGauge}. */
+public final class LongGaugeImpl extends LongGauge implements Meter {
+  private final MetricDescriptor metricDescriptor;
+  private volatile Map<List<LabelValue>, PointImpl> registeredPoints = Collections.emptyMap();
+  private final int labelKeysSize;
+  private final List<LabelValue> defaultLabelValues;
+
+  LongGaugeImpl(String name, String description, String unit, List<LabelKey> labelKeys) {
+    labelKeysSize = labelKeys.size();
+    defaultLabelValues = Collections.unmodifiableList(new ArrayList<LabelValue>(labelKeysSize));
```
Comment (bogdandrutu) [inline impl_core/src/main/java/io/opencensus/implcore/metrics/LongGaugeImpl.java]: This is not correct:
1. The `new ArrayList<LabelValue>(labelKeysSize)` creates an empty list with the capacity labelKeysSize. See https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html#ArrayList-int-
2. See https://github.com/census-instrumentation/opencensus-java/blob/master/api/src/main/java/io/opencensus/metrics/LabelValue.java. You should add LabelValues with null string to this list.

Please add a test for this.

File: impl_core/src/main/java/io/opencensus/implcore/metrics/LongGaugeImpl.java [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+  public LongPoint getOrCreateTimeSeries(List<LabelValue> labelValues) {
+    // lock free point retrieval, if it is present
+    PointImpl existingPoint = registeredPoints.get(labelValues);
+    if (existingPoint != null) {
+      return existingPoint;
+    }
+
+    checkNotNull(labelValues, "labelValues should not be null.");
+    checkArgument(labelKeysSize == labelValues.size(), "Incorrect number of labels.");
+    Utils.checkListElementNotNull(labelValues, "labelValues element should not be null.");
```
Comment (bogdandrutu) [inline impl_core/src/main/java/io/opencensus/implcore/metrics/LongGaugeImpl.java]: This is related to my previous comment. Consider to move this to the registerTimeSeries. Downside is first time when we call register for default we do an extra allocation for the Collections.unmodifiableList but we can amortize that by not calling Collections.unmodifiableList on the default in the constructor.

File: impl_core/src/test/java/io/opencensus/implcore/metrics/LongGaugeImplTest.java [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+  public void getOrCreateTimeSeries_WithLabels() {
+    LongPoint point = longGaugeMetric.getOrCreateTimeSeries(labelValues);
+    point.add(100);
+    point.add(-60);
+    point.set(500);
+
+    assertThat(longGaugeMetric.getMetric(testClock))
+        .isEqualTo(
+            Metric.create(
+                MetricDescriptor.create(
```
Comment (bogdandrutu) [inline impl_core/src/test/java/io/opencensus/implcore/metrics/LongGaugeImplTest.java]: This can be static variable.

File: impl_core/src/test/java/io/opencensus/implcore/metrics/LongGaugeImplTest.java [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+                    METRIC_NAME,
+                    METRIC_DESCRIPTION,
+                    METRIC_UNIT,
+                    Type.GAUGE_INT64,
+                    Collections.singletonList(LABEL_KEY)),
+                Collections.singletonList(
+                    TimeSeries.create(
+                        Collections.<LabelValue>emptyList(),
```
Comment (bogdandrutu) [inline impl_core/src/test/java/io/opencensus/implcore/metrics/LongGaugeImplTest.java]: This is incorrect. See comment in the ctor of the class why. default LabelValue does not mean missing value.

File: impl_core/src/test/java/io/opencensus/implcore/metrics/LongGaugeImplTest.java [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Unit tests for {@link LongGaugeImpl}. */
+@RunWith(JUnit4.class)
+public class LongGaugeImplTest {
```
Comment (bogdandrutu) [inline impl_core/src/test/java/io/opencensus/implcore/metrics/LongGaugeImplTest.java]: Consider to add tests for:
1. getOrCreateTimeSeries with default labels.
2. getOrCreateTimeSeries returns the same point if added multiple times (consider to create new lists to make sure that the equality is not by the reference).
3. Test remove (create / remove / create returns different point). After remove the getMetric return null (or no TimeSeries) 
4. Test clear - same as remove
5. Test all exceptions (nullpointer, different sizes).
6. Negative values for a gauge.

File: api/src/test/java/io/opencensus/metrics/MetricRegistryTest.java (hunk 10 above)
Comment (bogdandrutu) [inline api/src/test/java/io/opencensus/metrics/MetricRegistryTest.java]: Consider all tests related to the LongGauge (addPoint, remove, etc) to be moved in a LongGaugeTest class.

File: impl_core/src/test/java/io/opencensus/implcore/metrics/MetricRegistryImplTest.java [hunk 19]
```diff
@@ -53,248 +54,63 @@
...
-        null,
-        new ToDoubleFunction<Object>() {
-          @Override
-          public double applyAsDouble(Object value) {
-            return 5.0;
-          }
-        });
-  }
+  // helper class
+  public static class TotalMemory {
```
Comment (bogdandrutu) [inline impl_core/src/test/java/io/opencensus/implcore/metrics/MetricRegistryImplTest.java]: Not used.

File: impl_core/src/test/java/io/opencensus/implcore/metrics/MetricRegistryImplTest.java [hunk 20]
```diff
@@ -53,248 +54,63 @@
...
-                        Collections.unmodifiableList(Collections.singletonList(LABEL_VALUES)),
-                        Point.create(Value.doubleValue(5.0), TEST_TIME),
-                        null))));
+    public long getAvailable() {
+      return 7;
+    }
   }
 
   @Test
   public void addLongGauge_NullName() {
```
Comment (bogdandrutu) [inline impl_core/src/test/java/io/opencensus/implcore/metrics/MetricRegistryImplTest.java]: Add a test to make sure that you get a LongGaugeImpl. So check that the return object is instance of this class.

File: impl_core/src/test/java/io/opencensus/implcore/metrics/LongGaugeImplTest.java (hunk 18 above)
Comment (mayurkale22) [reply impl_core/src/test/java/io/opencensus/implcore/metrics/LongGaugeImplTest.java]: Done

File: impl_core/src/test/java/io/opencensus/implcore/metrics/MetricRegistryImplTest.java (hunk 20 above)
Comment (mayurkale22) [reply impl_core/src/test/java/io/opencensus/implcore/metrics/MetricRegistryImplTest.java]: Done

File: impl_core/src/main/java/io/opencensus/implcore/metrics/MetricRegistryImpl.java [hunk 21]
```diff
@@ -17,85 +17,71 @@
...
-      LinkedHashMap<LabelKey, LabelValue> labels,
-      T obj,
-      ToLongFunction<T> function) {
-    checkNotNull(labels, "labels");
-    registeredMeters.registerMeter(
-        new LongGauge<T>(
-            checkNotNull(name, "name"),
+  public LongGauge addLongGauge(
+      String name, String description, String unit, List<LabelKey> labelKeys) {
+    checkNotNull(name, "name");
```
Comment (bogdandrutu) [inline impl_core/src/main/java/io/opencensus/implcore/metrics/MetricRegistryImpl.java]: This is inconsistent with the Noop implementation. If you do checks in this class then do the checks in the MetricRegistry as well not in the NoopLongGauge.

File: impl_core/src/main/java/io/opencensus/implcore/metrics/MetricRegistryImpl.java [hunk 22]
```diff
@@ -17,85 +17,71 @@
...
 import java.util.LinkedHashSet;
+import java.util.List;
 import java.util.Set;
+import javax.annotation.Nullable;
 
 /** Implementation of {@link MetricRegistry}. */
 public final class MetricRegistryImpl extends MetricRegistry {
   private final RegisteredMeters registeredMeters;
   private final MetricProducer metricProducer;
+  private volatile Set<String> registeredMetrics = Collections.emptySet();
```
Comment (bogdandrutu) [inline impl_core/src/main/java/io/opencensus/implcore/metrics/MetricRegistryImpl.java]: Why do we need to sets (name, Meters) instead of a map in the RegisteredMeters

File: api/src/main/java/io/opencensus/metrics/LongGauge.java (hunk 8 above)
Comment (mayurkale22) [reply api/src/main/java/io/opencensus/metrics/LongGauge.java]: Done

File: impl_core/src/main/java/io/opencensus/implcore/internal/Utils.java (hunk 13 above)
Comment (mayurkale22) [reply impl_core/src/main/java/io/opencensus/implcore/internal/Utils.java]: Somehow failed to push in previous commit, added now.

File: impl_core/src/main/java/io/opencensus/implcore/metrics/Meter.java [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+
+interface Meter {
+  /**
+   * Provides a {@link io.opencensus.metrics.export.Metric} with one or more {@link
+   * io.opencensus.metrics.export.TimeSeries}.
+   *
+   * @param clock the clock used to get the time.
+   * @return a {@code Metric}.
+   */
+  @Nullable
```
[the remark this answers is not in the document]
Comment (mayurkale22) [reply impl_core/src/main/java/io/opencensus/implcore/metrics/Meter.java]: Done

**Comment by mayurkale22** [comment]
@bogdandrutu This PR contains only LongGauge/NoopLongGauge and LongGaugeImpl. PTAL

File: impl_core/src/main/java/io/opencensus/implcore/metrics/LongGaugeImpl.java [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+    return newPoint;
+  }
+
+  @Nullable
+  @Override
+  public Metric getMetric(Clock clock) {
+    Map<List<LabelValue>, PointImpl> currentRegisteredPoints = registeredPoints;
+
+    int pointCount = currentRegisteredPoints.size();
+    if (pointCount > 0) {
```
Comment (bogdandrutu) [inline impl_core/src/main/java/io/opencensus/implcore/metrics/LongGaugeImpl.java]: Add a TODO to say optimize this for 1 timeseries. Not more than that.

File: impl_core/src/main/java/io/opencensus/implcore/metrics/LongGaugeImpl.java [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+    if (existingPoint != null) {
+      return existingPoint;
+    }
+    return registerTimeSeries(defaultLabelValues);
+  }
+
+  @Override
+  public synchronized void removeTimeSeries(List<LabelValue> labelValues) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+    Utils.checkListElementNotNull(labelValues, "labelValues element should not be null.");
```
Comment (bogdandrutu) [inline impl_core/src/main/java/io/opencensus/implcore/metrics/LongGaugeImpl.java]: this line and the next one are not need probably for remove.
